### PR TITLE
grpc: addpsbtoutput command

### PR DESCRIPTION
**Pull request by s373nZ**
@cdecker This PR is a first pass at #6986. The major fix here was including `AddPsbtOutput` to the big list of RPC commands in `utils.py` for `msggen`. After re-running the `bundle` and `generate` commands, custom changes to the `.msggen.json` file were necessary to map fields and set versions. The rest of the files were generated automatically.

Also added parameter descriptions and some references to `addpsbtoutput` in the docs where it seemed to make sense in a separate commit.

I was a little paranoid about committing the generated code in case there were small differences between build environments, and ended up double-checking by building with Docker. Hope it worked out...

@BitcoinJiuJitsu I'm hoping this qualifies to claim the Sphinx bounty, too ;)

### CODE REVIEW

**Comment by s373nZ** [comment]
@cdecker Pretty sure this CI got disrupted by the Poetry `1.8.0` release a few hours ago. The `Dockerfile.alpine` works locally for me when I change the command to `python3 -m pip install "poetry==1.7.1"`, but I imagine this might break elsewhere though...

**Comment by s373nZ** [comment]
Trying to avoid errors in the CI Prebuild check [here](https://github.com/ElementsProject/lightning/actions/runs/8038975834/job/21955802811) by running a build using: `docker build -t lightningd-ubuntu -f contrib/docker/Dockerfile.ubuntu .`

And committing the generated code from that.

**Comment by cdecker** [comment]
> @cdecker Pretty sure this CI got disrupted by the Poetry `1.8.0` release a few hours ago. The `Dockerfile.alpine` works locally for me when I change the command to `python3 -m pip install "poetry==1.7.1"`, but I imagine this might break elsewhere though...

I'm removing the bloody alpine test, it keeps rotting away. See #7114. This PR will not make it into the v24.02 anyway, the feature freeze was 2 weeks ago :-)

**Comment by s373nZ** [comment]
> I'm removing the bloody alpine test, it keeps rotting away. See #7114. This PR will not make it into the v24.02 anyway, the feature freeze was 2 weeks ago :-)

I removed the Python pin commit for Alpine then. I don't mean to waste build cycles or attention on review notifications -- if there is a standard way to run the build, codgen and doc tests locally, please do share. So far I've been using:
```
./configure
make
make check
```
But the `check` reports some different command names (even on `master`) so I'm not sure how far it gets. Also been experimenting trying to run the various builds through the `Dockerfiles`... Currently running the above under that experimental Nix flake, so there is ample room for confusion :/

**Comment by cdecker** [comment]
This being a `msggen`-related change, you probably want to `make check-gen-updated` which is what is run on CI to check that all dependent files have beenm updated, and then `make pytest` to run through the python integration tests.

**Comment by s373nZ** [comment]
@cdecker I rebased the latest from `master`. After experiencing some inconsistent behavior from `pytest` locally, I managed to produce two successful consecutive runs after:
1. Freeing up some more disk space
2. Raising the `ulimit` for `open files` as I saw some related crashes when `pyln` was reading the `bitcoin.conf`

In any event, I'm happy to have a development environment working as expected. Thanks for fielding questions in Discord. AFAICT, it seems like the next step is another round of CI.

**Comment by cdecker** [comment]
Hm, that `ulimit` comment is concerning, we should never have more than 1024 FDs open at a time, and this might indicate an FD leak.

**Comment by s373nZ** [comment]
@cdecker I ran another `make pytest` locally with `ulimit` at the default of `1024` and was able to reproduce the `Too many open files` error. Yes, there is an FD leak and I think I've narrowed it down to:

**Comment by s373nZ** [comment]
@cdecker I re-ran `pytest` today with `ulimit` set to `1024` and yes, there is an FD leak. It shows up under my Nix setup, because Nix also opens a lot of files [1](https://github.com/NixOS/nix/issues/6007) [2](https://github.com/NixOS/nix/issues/10158). I think I've narrowed it down to this:

https://github.com/ElementsProject/lightning/blob/d1101f416f1ec959981a8ebe5639a6b267885bfd/contrib/pyln-testing/pyln/testing/utils.py#L203

Just added commit `[42e1e6b](https://github.com/ElementsProject/lightning/pull/7108/commits/42e1e6bd470db41c5a6ed9b91d54640e87730a0e)` which seems to resolve the FD leak for me locally. A few notes:
1. Initially I tried to do this in a `__del__()` destructor, but it didn't work -- maybe because `NodeFactory` is a fixture.
2. Although `stop()` calls `kill()` in some cases, if there are tests that call `kill()` directly, maybe there is a more robust place to centralize the cleanup.
3. Let me know if you would like to open a new issue or cherry-pick to a new PR for the FD leak topic, as it's not specifically related to the `addpsbtoutput` command.

**Comment by s373nZ** [comment]
Removed the FD leak changes due to a few failing tests and opened https://github.com/ElementsProject/lightning/issues/7130 to investigate them there.

**Comment by daywalker90** [comment]
This is still based on the old json schema format. Will you update this PR @s373nZ ?

**Comment by daywalker90** [comment]
IIRC you should not change the `added` version but instead use the patch.py to backfill the fields:
- uncomment these lines in `contrib/msggen/msggen/patch.py`:
```python
# if f.added is None and 'added' not in m:
#    m['added'] = 'pre-v0.10.1'
```
- run msggen
- comment lines in patch.py again
- commit

**Comment by s373nZ** [comment]
@daywalker90 I did as you suggested. Still a little confused why the `satoshi` field isn't included in the `.msggen.json` `model-field-versions`, but it is of type `sat` and the other example of that field in `doc/schemas/lightning-funderupdate.json` named `policy_mod` isn't included either. Perhaps it is excluded as a complex field.

Thanks for the feedback. Let me know if you see anything else.

**Comment by daywalker90** [comment]
Ah yes you have to change the `sat` type to `msat` for now. Then rerun msggen

**Comment by s373nZ** [comment]
@daywalker90 Nice -- that looks better to me, although I don't know much of what to expect from the generated code. Also rebased against `master`.

**Comment by daywalker90** [comment]
looks good to me

**Comment by cdecker** [comment]
Rebased on top of `master`. Hopefully we can make it into the release, no reason not to try :-)

Sorry for forgetting about this one :frowning_face:

**Comment by s373nZ** [comment]
Thanks! It had almost slipped on me as well. I've submitted a plea for inclusion over Discord.

File: contrib/msggen/msggen/schema.json [hunk 1]
```diff
@@ -79,7 +79,7 @@
         ],
         "properties": {
           "satoshi": {
-            "type": "sat",
+            "type": "msat",
```
Comment (endothermicdev) [inline contrib/msggen/msggen/schema.json:82]: This doesn't look right.  The psbt code handles satoshi amounts and the description matches this.  Maybe there needs to be a new rust primitive type?

Comment (daywalker90) [reply contrib/msggen/msggen/schema.json:82]: see here: https://github.com/ElementsProject/lightning/pull/7230#discussion_r1571096707

Comment (endothermicdev) [reply contrib/msggen/msggen/schema.json:82]: Ah, thanks for the back story!

**Comment by ShahanaFarooqui** [comment]
ACK 619687193859195a56aab97c0af5c0002c08e3b9.

**Pull request merged**